driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/gridfs/GridFSUploadPublisherImpl.java

```diff
@@ -149,38 +149,15 @@ public void subscribe(final Subscriber<? super ObjectId> subscriber) {
     }
 
     private Mono<Void> createCheckAndCreateIndexesMono(@Nullable final Timeout timeout) {
-        AtomicBoolean collectionExists = new AtomicBoolean(false);
-        return Mono.create(sink -> findAllInCollection(filesCollection, timeout).subscribe(
-                        d -> collectionExists.set(true),
-                        sink::error,
-                        () -> {
-                            if (collectionExists.get()) {
-                                sink.success();
-                            } else {
-                                checkAndCreateIndex(filesCollection.withReadPreference(primary()), FILES_INDEX, timeout)
-                                        .doOnSuccess(i -> checkAndCreateIndex(chunksCollection.withReadPreference(primary()), CHUNKS_INDEX, timeout)
-                                                .subscribe(unused -> {}, sink::error, sink::success))
-                                        .subscribe(unused -> {}, sink::error);
-                            }
-                        })
-        );
-    }
-
-    private Mono<Document> findAllInCollection(final MongoCollection<GridFSFile> collection, @Nullable final Timeout timeout) {
-        return collectionWithTimeoutDeferred(collection
-                .withDocumentClass(Document.class)
-                .withReadPreference(primary()), timeout)
-                .flatMap(wrappedCollection -> {
-                    if (clientSession != null) {
-                        return Mono.from(wrappedCollection.find(clientSession)
-                                .projection(PROJECTION)
-                                .first());
-                    } else {
-                        return Mono.from(wrappedCollection.find()
-                                .projection(PROJECTION)
-                                .first());
-                    }
-                });
+        return collectionWithTimeoutDeferred(filesCollection.withDocumentClass(Document.class).withReadPreference(primary()), timeout)
+                .map(collection -> clientSession != null ? collection.find(clientSession) : collection.find())
+                .flatMap(findPublisher -> Mono.from(findPublisher.projection(PROJECTION).first()))
+                .switchIfEmpty(Mono.defer(() ->
+                        checkAndCreateIndex(filesCollection.withReadPreference(primary()), FILES_INDEX, timeout)
+                                .then(checkAndCreateIndex(chunksCollection.withReadPreference(primary()), CHUNKS_INDEX, timeout))
+                                .then(Mono.fromCallable(Document::new))
```

Comment:
It seems we can easily avoid having `Mono.fromCallable(Document::new)` by flat mapping `Mono<Document>` to `Mono<Void>` via `then` before we call `switchIfEmpty`:

```suggestion
                .flatMap(findPublisher -> Mono.from(findPublisher.projection(PROJECTION).first()))
                .then()
                .switchIfEmpty(Mono.defer(() ->
                        checkAndCreateIndex(filesCollection.withReadPreference(primary()), FILES_INDEX, timeout)
                                .then(checkAndCreateIndex(chunksCollection.withReadPreference(primary()), CHUNKS_INDEX, timeout))
                                .then()
```

Reply:
As `then()` will never have a result `switchIfEmpty` will always be called.

> Mono<Void> should be used for Publisher that just completes without any value.
https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Mono.html

Reply:
Updated to use `.then(Mono.empty())` instead of `.then(Mono.fromCallable(Document::new))`

Reply:
> As `then()` will never have a result `switchIfEmpty` will always be called.

This was a stupid suggestion from me. Sorry, Ross.